### PR TITLE
Rewrite description of `bool`'s `should`

### LINKED ISSUE #23831: Bool minimum_should_match with "should" clause clarity
The [bool query docs](https://www.elastic.co/guide/en/elasticsearch/reference/5.2/query-dsl-bool-query.html) aren't clear - it does say " In a boolean query with no must or filter clauses, one or more should clauses must match a document. ", but what it doesn't say is that when there IS a must or filter clause, then 0 should clauses must match.   

Recommended adding something (to the _should_ clause section) like:

> When a must or filter clause is present, minimum_should_match defaults to 0.

### PR DESCRIPTION
**Pull request by nik9000**
Rewrites the description of the `bool` query's `should`
clauses so it is (hopefully) more clear what the defaults
for `minimum_should_match` are.

There is still an `[IMPORTANT]` section about `minimum_should_match`
in a filter context. I think it is worth keeping because it is, well,
important.

Closes #23831

### CODE REVIEW

File: docs/reference/query-dsl/bool-query.asciidoc [hunk 1]
```diff
@@ -17,12 +17,15 @@ contribute to the score.
...
-must match a document. The minimum number of should clauses to match can
-be set using the
-<<query-dsl-minimum-should-match,`minimum_should_match`>>
-parameter.
+|`should` |The clause (query) should appear in the matching document. If the
+`bool` query is in a <<query-filter-context,query context>> and has a `must` or
+`filter` clause then a document will match the `bool` query even if none of the
+`should` queries match. In this case these clauses are only used to influence
+the score. If the `bool` query is a <<query-filter-context,filter context>>
+or has neither `must` or `should` then at least one of the `should` queries
```
Comment (jimczi) [inline docs/reference/query-dsl/bool-query.asciidoc]: `must` or `filter` ?

**Review by jimczi — APPROVED** [review]
LGTM

File: docs/reference/query-dsl/bool-query.asciidoc (hunk 1 above)
Comment (nik9000) [inline docs/reference/query-dsl/bool-query.asciidoc]: Right! Thanks that.

**Comment by nik9000** [comment]
Thanks for reviewing @jimczi!

**Pull request merged**